examples/instances_reserved_public_ip/variables.tf

```diff
@@ -0,0 +1,131 @@
+# Copyright (c) 2019, 2021, Oracle Corporation and/or affiliates.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl
+
+# provider identity parameters
+variable "fingerprint" {
+  description = "fingerprint of oci api private key"
+  type        = string
+  # no default value, asking user to explicitly set this variable's value. see codingconventions.adoc
+}
+
+variable "private_key_path" {
+  description = "path to oci api private key used"
+  type        = string
+  # no default value, asking user to explicitly set this variable's value. see codingconventions.adoc
+}
+
+variable "region" {
+  description = "the oci region where resources will be created"
+  type        = string
+  # no default value, asking user to explicitly set this variable's value. see codingconventions.adoc
+  # List of regions: https://docs.cloud.oracle.com/iaas/Content/General/Concepts/regions.htm#ServiceAvailabilityAcrossRegions
+}
+
+variable "tenancy_ocid" {
+  description = "tenancy ocid where to create the sources"
+  type        = string
+  # no default value, asking user to explicitly set this variable's value. see codingconventions.adoc
+}
+
+variable "user_ocid" {
+  description = "ocid of user that terraform will use to create the resources"
+  type        = string
+  # no default value, asking user to explicitly set this variable's value. see codingconventions.adoc
+}
+
+# general oci parameters
+
+variable "compartment_ocid" {
+  description = "compartment ocid where to create all resources"
+  type        = string
+  # no default value, asking user to explicitly set this variable's value. see codingconventions.adoc
+}
+
+variable "freeform_tags" {
+  description = "simple key-value pairs to tag the resources created using freeform tags."
+  type        = map(string)
+  default     = null
+}
+
+variable "defined_tags" {
+  description = "predefined and scoped to a namespace to tag the resources created using defined tags."
+  type        = map(string)
+  default     = null
+}
+
+# compute instance parameters
+
+variable "instance_ad_number" {
+  description = "The availability domain number of the instance. If none is provided, it will start with AD-1 and continue in round-robin."
+  type        = number
+  default     = 1
+}
+
+variable "instance_count" {
+  description = "Number of identical instances to launch from a single module."
+  type        = number
+  default     = 1
+}
+
+variable "instance_display_name" {
+  description = "(Updatable) A user-friendly name for the instance. Does not have to be unique, and it's changeable."
+  type        = string
+  default     = "module_instance_flex"
+}
+
+variable "instance_flex_memory_in_gbs" {
+  type        = number
+  description = "(Updatable) The total amount of memory available to the instance, in gigabytes."
+  default     = null
```

Comment:
Please set default value to that which OCI will automatically assign. Null should be used if we cannot assign a value or do not want to assign a value. The 1st case is to be used in conditionals so if something is null, don't do this. The 2nd case is to make it mandatory for the user to provide these values.

Reply:
this is set to `null` because it handle both flex and fixed shapes with a single pair of variables:
- it is used in conditionals to fall-back from flex-shape to fixed-shape
- it must be optional even with flex shapes, the default cpu/mem values can be retrieved and applied 

There is lot of logic behind it that came with PR #49.